### ISSUE TITLE
MSBuild Errors while building Kudu with BuildTools for Visual Studio 2017

### ISSUE THREAD

**Issue by sabeersulaiman** [projectkudu/kudu#2542]
I have cloned Kudu to a Windows Server 2012, and installed BuildTools for Visual Studio 2017. When I try to build using the ```build.cmd``` command it fails with the following errors:

```
Microsoft (R) Build Engine version 4.7.2053.0
[Microsoft .NET Framework, version 4.0.30319.42000]
Copyright (C) Microsoft Corporation. All rights reserved.

  MSBuild auto-detection: using msbuild version '15.3.409.57025' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\bin'.
  All packages listed in packages.config are already installed.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
C:\Users\sabeer\Downloads\kudu\Kudu.Services.Web\Kudu.Services.Web.csproj(267,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v11.0\WebApplications\Microsoft.WebApplication.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
Functions\FunctionSecretsJsonOps.cs(23,24): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\FunctionSecretsJsonOps.cs(27,24): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\FunctionSecretsJsonOps.cs(66,39): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\FunctionTestData.cs(9,54): error CS1519: Invalid token '=' in class, struct, or interface member declaration [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\FunctionTestData.cs(9,77): error CS1519: Invalid token ';' in class, struct, or interface member declaration [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\MasterKeyJsonOps.cs(22,24): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\MasterKeyJsonOps.cs(26,24): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
Functions\MasterKeyJsonOps.cs(51,39): error CS1056: Unexpected character '$' [C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\Kudu.Contracts.csproj]
```

What I noticed is that I have MSBuild tools in ```C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild``` But it seems the ```C:\Program Files (x86)\MSBuild``` folder is automatically used. Both only contains ```v15.0``` folder and not ```v11.0``` folder specified in the error. If there are any missing MSBuild Packages, where can I download them?

**Comment by davidebbo** [comment]
Interesting. So on a machine with full VS 2017 it builds fine, but when installing just the build tools it doesn't?

**Comment by watashiSHUN** [comment]
not sure if I understand your question, are you trying to build kudu using msbuild15?
based on your error message:

> MSBuild auto-detection: using msbuild version '15.3.409.57025' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\bin'.

it looks like `nuget` was able to discover your BuildTool installed,
I am not sure whether `build.cmd` will find this BuildTool....since it's not in the discovery path:
https://github.com/projectkudu/kudu/blob/afa47eb71d5ac7b72a6c4c27cc7b222b4ff03479/build.cmd#L10-L25

the error message probably confirms it
> Microsoft (R) Build Engine version 4.7.2053.0
> [Microsoft .NET Framework, version 4.0.30319.42000]
> Copyright (C) Microsoft Corporation. All rights reserved.

since this is the same as 
https://github.com/projectkudu/kudu/blob/afa47eb71d5ac7b72a6c4c27cc7b222b4ff03479/build.cmd#L8
you can try to add this to the top ?
```
if exist "%PROGRAMFILES%\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\bin\MSBuild.exe" ( 
     set MsBuildExe="%PROGRAMFILES%\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\bin\MSBuild.exe" 
```

**Comment by sabeersulaiman** [comment]
As @watashiSHUN pointed out, not detecting build tools versions of MSBuild was the case for this error it seems. Adding the new discovery path as @watashiSHUN recommended solved that issue. Now I am getting some other errors. Now trying to figure this out

```
Microsoft (R) Build Engine version 15.3.409.57025 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  MSBuild auto-detection: using msbuild version '15.3.409.57025' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\bin'.
  All packages listed in packages.config are already installed.
  Kudu.Contracts -> C:\Users\sabeer\Downloads\kudu\Kudu.Contracts\bin\Release\Kudu.Contracts.dll
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1987,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "System.Web.Http". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\Users\sabeer\Downloads\kudu\Kudu.Core\Kudu.Core.csproj]
  Kudu.Core -> C:\Users\sabeer\Downloads\kudu\Kudu.Core\bin\Release\Kudu.Core.dll
  Kudu.Services -> C:\Users\sabeer\Downloads\kudu\Kudu.Services\bin\Release\Kudu.Services.dll
  Kudu.Console -> C:\Users\sabeer\Downloads\kudu\Kudu.Console\bin\Release\kudu.exe
  Info: No event source classes needing registration found in C:\Users\sabeer\Downloads\kudu\Kudu.Console\bin\Release\kudu.exe
  Kudu.Client -> C:\Users\sabeer\Downloads\kudu\Kudu.Client\bin\Release\Kudu.Client.dll
  Kudu.Services.Web -> C:\Users\sabeer\Downloads\kudu\Kudu.Services.Web\bin\Kudu.Services.Web.dll
  Info: No event source classes needing registration found in C:\Users\sabeer\Downloads\kudu\Kudu.Services.Web\bin\Kudu.Services.Web.dll
  Running post build events
  Kudu.SiteManagement -> C:\Users\sabeer\Downloads\kudu\Kudu.SiteManagement\bin\Release\Kudu.SiteManagement.dll
  Kudu.TestHarness -> C:\Users\sabeer\Downloads\kudu\Kudu.TestHarness\bin\Release\Kudu.TestHarness.dll
  Kudu.Core.Test -> C:\Users\sabeer\Downloads\kudu\Kudu.Core.Test\bin\Release\Kudu.Core.Test.dll
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1987,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Microsoft.VisualStudio.QualityTools.LoadTestFramework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1987,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Microsoft.VisualStudio.QualityTools.UnitTestFramework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1987,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Microsoft.VisualStudio.QualityTools.WebTestFramework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
  Kudu.Services.Test -> C:\Users\sabeer\Downloads\kudu\Kudu.Services.Test\bin\Release\Kudu.Services.Test.dll
KuduStress.cs(13,17): error CS0234: The type or namespace name 'VisualStudio' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
StressTestCases.cs(7,17): error CS0234: The type or namespace name 'VisualStudio' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
StressTestCases.cs(45,10): error CS0246: The type or namespace name 'TestMethodAttribute' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
StressTestCases.cs(45,10): error CS0246: The type or namespace name 'TestMethod' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
StressTestCases.cs(59,10): error CS0246: The type or namespace name 'TestMethodAttribute' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
StressTestCases.cs(59,10): error CS0246: The type or namespace name 'TestMethod' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\sabeer\Downloads\kudu\Kudu.Stress\Kudu.Stress.csproj]
  Kudu.Performance -> C:\Users\sabeer\Downloads\kudu\Kudu.Performance\bin\Release\Kudu.Performance.exe
  Kudu.SiteManagement.Test -> C:\Users\sabeer\Downloads\kudu\Kudu.SiteManagement.Test\bin\Release\Kudu.SiteManagement.Test.dll
  Kudu.FunctionalTests -> C:\Users\sabeer\Downloads\kudu\Kudu.FunctionalTests\bin\Release\Kudu.FunctionalTests.dll
```

**Comment by sabeersulaiman** [comment]
Is it best to just build kudu from a development machine and install it onto the server. What is the recommended method, it doesn't seem like I can get the build to work without installing VS. It seems I am missing some test frameworks. I am new to MSBuild and Windows Build System in general.

**Comment by sabeersulaiman** [comment]
I have managed to build Kudu on my dev machine and set it up on the production server. Created an application. As I understand the git repository is at ```http://localhost:88966/app_name.git```. I was wondering how to securely setup continuous deployment from my remote git to this server. I noticed that Kudu service URL is also on the same port. Can someone point me in the right direction as to set this up in a secure way.
